src/earthkit/hydro/readers.py

```
@@ -1,224 +1,115 @@
+                os.makedirs(cache_dir, exist_ok=True)
+        else:
+            print("Cache disabled.")
 
-    return create_network(upstream_indices, downstream_indices, missing_mask, shape)
+        network = func(path, river_format, source)
 
+        if use_cache:
+            joblib.dump(network, cache_filepath, compress=cache_compression)
+            print(f"River network loaded, saving to cache ({cache_filepath}).")
 
-def from_netcdf_cama(filename, type="nextxy"):
-    """
-    Loads a river network from a CaMa-Flood style NetCDF file.
+        return network
 
-    Parameters
-    ----------
-    filename : str
-        Path to the NetCDF file.
-    type : str, optional
-        Type of CaMa-Flood encoding to use ("downxy" or "nextxy", default is "downxy").
+    return wrapper
 
-    Returns
-    -------
-    RiverNetwork
-        The constructed river network object.
 
-    Raises
-    ------
-    Exception
-        If an unknown type is specified.
-    """
+def import_earthkit_or_prompt_install(river_format, source):
     try:
         import earthkit.data as ekd
     except ModuleNotFoundError:
         raise ModuleNotFoundError(
-            "earthkit-data is required for netcdf support.\nTo install it, run `pip install earthkit-data`"
+            f"earthkit-data is required for loading {river_format} from {source}.\nTo install it, run `pip install earthkit-data`"
         )
-    data = ekd.from_source("file", filename).to_xarray(mask_and_scale=False)
-    if type == "downxy":
-        dx, dy = data.downx.values, data.downy.values
-        return from_cama_downxy(dx, dy)
-    elif type == "nextxy":
+    return ekd
+
+
+@cache
+def create_river_network(path, river_format, source):
+    if river_format == "precomputed":
+        if source == "file":
+            return joblib.load(path)
+        elif source == "url":
+            return joblib.load(BytesIO(urlopen(path).read()))
+        else:
```

Shouldn't we return an error instead here? What would be the case for else?

---

The else would allow one to load the network pickle from any other ekd-supported sources e.g. ECFS

Can raise an error if we don't want to support this

---

I don't think this will be needed. URL and file should already be more than enough.

---

Addressed in 7d8e630743dbf96a6ea85877c1337c25d4528bfc